Progress Bar component

claudiulodro opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Closes #35  .

Adds a Muriel progress bar component. I've followed the Muriel style for this instead of the rounded style in the screenshot in #32.

<img width="740" alt="Screen Shot 2019-05-09 at 12 02 22 PM" src="https://user-images.githubusercontent.com/7317227/57479492-64784880-7252-11e9-9b87-8d7e7aba00da.png">

### How to test the changes in this Pull Request:

1. `npm ci` `npm run build:webpack` `npm run test`
2. Go to Subscriptions page and observe the varieties of progress bar available:
<img width="472" alt="Screen Shot 2019-05-09 at 12 03 45 PM" src="https://user-images.githubusercontent.com/7317227/57479563-8e316f80-7252-11e9-80e3-c017b81fdcfd.png">

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

jeffersonrabb commented:
Note: rebase: https://github.com/Automattic/newspack-plugin/compare/37910cb0f0102b2186277e58aba4f51f6c4b1194..db7c7c88e8c407abe8e60d48510bad3ae8ce232d

jeffersonrabb commented on assets/src/components/progressBar/index.js:
```diff
@@ -0,0 +1,70 @@
+	 *
+	 * @return int
+	 */
+	getCompletionPercentage() {
```
Suggestion: instead of getting these values from props rewrite the function to take two parameters (completed, total) and return a value. It would be more easily testable and allows for sanitization to occur in only one place (more on this in a separate comment).

jeffersonrabb commented on assets/src/components/progressBar/index.js:
```diff
@@ -0,0 +1,70 @@
+	 */
+	render() {
+		const { label, completed, total, displayFraction } = this.props;
+
```
Suggestion: sanitize the completed value to check that it is a number and is between 0-`total`, e.g.:

`const cleanCompleted = Math.max( 0, Math.min( ( parseInt( completed ) || 0 ), parseInt( total ) ) );`

In my testing it's possible to set the `completed` prop to a number greater than total, a number less than 0, or a string that yields `NaN` from `parseInt()`.

jeffersonrabb commented on assets/src/components/progressBar/style.scss:
```diff
@@ -0,0 +1,36 @@
+.muriel-progress-bar {
+	margin: 16px 0;
+	position: relative;
+	z-index: 1;
```
This is kind of a fly-by, but given our aspirations to publish these components as a package for other consumers this might be a good moment to discuss `z-index` strategy. `z-index` normalization seems to be a major pain point for packaging Calypso components, since the mixin that handles  them works across the entirety of Calypso styles, which returns values which might be completely inappropriate in a different consumer App. Since this is (I believe) the first `z-index` in our component work, it would be great to strategize.

claudiulodro replied:
[From here:](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context)
> The z-index values of its child stacking contexts only have meaning in [the] parent.

As far as I can tell, we should be OK in all situations here unless e.g. the wizard's stylesheet overrides the z-indexing of this component. In this specific situation, the z-index is required so that the gray part of the bar is placed under the pink part of the bar.

Can you provide an example of the issues you describe? I'm having a hard time coming up with a strategy without knowing what issues to avoid.

jeffersonrabb commented on assets/src/components/progressBar/style.scss:
```diff
@@ -0,0 +1,36 @@
+
+	.muriel-progress-bar__bar {
+		height: 4px;
+		background-color: $muriel-hot-pink-500;
```
I wonder if a red-ish color like this tracks as indicating errors rather than progress? Let's revisit with Sonja, but I'm wondering if it should be something more  like `$muriel-blue-400`

jeffersonrabb commented on assets/src/wizards/subscriptions/index.js:
```diff
@@ -92,6 +93,15 @@ class SubscriptionsWizard extends Component {
+					/>
+					<ProgressBar completed="2" total="3" />
+					<ProgressBar completed="2" total="5" label="Progress made" />
+					<ProgressBar completed="0" total="5" displayFraction />
```
Not suggesting this be committed, but as a way to test this component I set the `completed` attribute to `{ inputTextValue2 }`, which lets me change progress in real time by changing the value of the second text input.

jeffersonrabb commented on assets/src/components/progressBar/style.scss:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	.muriel-progress-bar__bar {
+		height: 4px;
```
For bonus points we could animate the progress bar?

```
-webkit-transition: width 0.25s; /* Safari */
  transition: width 0.25s;
```

claudiulodro commented:
I believe the bar design with the steps in it is supposed to be something similar to this:

<img width="778" alt="Screen Shot 2019-05-10 at 10 36 42 AM" src="https://user-images.githubusercontent.com/7317227/57545601-b4691500-730f-11e9-925d-433799f6914d.png">

I'm unclear whether it should fill up like a progress bar, or if only the current wizard screen should be the highlighted chunk of that bar.

If it's not a big deal I'd prefer to use the bar in this PR for everything, though. :) That would likely just mean adding support for multiple colors to this bar.

claudiulodro commented:
I've made the requested revisions:
- Isolated `getCompletionPercentage` and added some tests.
- Cleaned variables better in `render`.
- Animated progress bar (looks cool 👍)

I don't believe the z-index here requires a greater z-index strategy, and we can follow up with any bar color issues in a separate PR.
